modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/BulkCopyEntriesMVCActionCommand.java

```diff
@@ -0,0 +1,213 @@
+
+			if (errorCount < 10) {
+				errorMessage = themeDisplay.translate(
+					"the-following-documents-folders-could-not-be-copied");
+			}
+
+			JSONPortletResponseUtil.writeJSON(
+				actionRequest, actionResponse,
+				JSONUtil.put("errorMessage", errorMessage));
+
+			hideDefaultSuccessMessage(actionRequest);
+		}
+		else {
+			JSONPortletResponseUtil.writeJSON(
+				actionRequest, actionResponse, _jsonFactory.createJSONObject());
+		}
+	}
+
+	private void _checkDestinationGroup(Group group) throws Exception {
+		if ((group != null) && group.isStaged() && !group.isStagingGroup()) {
+			throw new PortalException(
+				"cannot-copy-into-the-live-version-of-a-group");
+		}
+	}
+
+	private int _copyEntries(ActionRequest actionRequest) throws Exception {
+		int errorCount = 0;
+
+		long destinationFolderId = ParamUtil.getLong(
+			actionRequest, "destinationFolderId");
+
+		long destinationRepositoryId = ParamUtil.getLong(
+			actionRequest, "destinationRepositoryId");
+
+		Group group = _groupLocalService.fetchGroup(destinationRepositoryId);
+
+		_checkDestinationGroup(group);
+
+		long[] currentAndAncestorSiteAndDepotGroupIds =
+			_siteConnectedGroupGroupProvider.
+				getCurrentAndAncestorSiteAndDepotGroupIds(group.getGroupId());
+
+		long[] entryIds = ParamUtil.getLongValues(
+			actionRequest, "selectedEntries");
+
+		for (long entryId : entryIds) {
+			try {
+				DLFileEntry dlFileEntry =
+					_dlFileEntryLocalService.fetchDLFileEntry(entryId);
+
+				if (dlFileEntry != null) {
+					_dlAppService.copyFileEntry(
+						dlFileEntry.getFileEntryId(), destinationFolderId,
+						destinationRepositoryId,
+						dlFileEntry.getFileEntryTypeId(),
+						currentAndAncestorSiteAndDepotGroupIds,
+						ServiceContextFactory.getInstance(
+							DLFileEntry.class.getName(), actionRequest));
+
+					continue;
+				}
+
+				DLFileShortcut dlFileShortcut =
+					_dlFileShortcutLocalService.fetchDLFileShortcut(entryId);
+
+				if (dlFileShortcut != null) {
+					_dlAppService.copyFileShortcut(
+						dlFileShortcut.getFileShortcutId(), destinationFolderId,
+						destinationRepositoryId,
+						ServiceContextFactory.getInstance(
+							DLFileShortcut.class.getName(), actionRequest));
+
+					continue;
+				}
+
+				DLFolder dlFolder = _dlFolderLocalService.fetchDLFolder(
+					entryId);
+
+				if (dlFolder != null) {
+					_dlAppService.copyFolder(
+						dlFolder.getRepositoryId(), dlFolder.getFolderId(),
+						destinationRepositoryId, destinationFolderId,
+						_getFileEntryTypeIds(
+							group.getGroupId(), dlFolder.getFolderId()),
+						currentAndAncestorSiteAndDepotGroupIds,
+						ServiceContextFactory.getInstance(
+							DLFolder.class.getName(), actionRequest));
+				}
+			}
+			catch (PortalException portalException) {
+				if (_log.isDebugEnabled()) {
+					_log.debug(portalException);
+				}
+
+				errorCount++;
+			}
+		}
+
+		return errorCount;
+	}
+
+	private Map<Long, Long> _getFileEntryTypeIds(long groupId, long folderId)
```

Comment:
This method and the method `_checkDestinationGroup` are shared with the other copy ActionCommand. Is it worth to have a Base ActionCommand for reusing of code?